[Functions] Abstract repetitive code and add unit tests. #9502

golden-yang opened the pull request:
<!--
### Contribution Checklist
  - Name the pull request in the form "[Issue XYZ][component] Title of the pull request", where *XYZ* should be replaced by the actual issue number.
    Skip *Issue XYZ* if there is no associated github issue for this pull request.
    Skip *component* if you are unsure about which is the best component. E.g. `[docs] Fix typo in produce method`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.

**(The sections below can be removed for hotfixes of typos)**
-->


### Motivation
This PR is related to #9502 and [PR #9519](https://github.com/apache/pulsar/pull/9519).
In [PR #9519 ](https://github.com/apache/pulsar/pull/9519), another contributer has fixed the problem described in #9502, but there are still some areas that can be optimized. 

### Modifications
- Abstract a method and then reuse some code.
- Adding a unit test for the `restartFunctionInstances` method


And I am a newcomer to the pulsar community, do I need to recreate an issue?

eolivelli commented on pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:
```diff
@@ -452,25 +444,28 @@ public synchronized void restartFunctionInstances(String tenant, String namespac
         return;
     }
 
+    private void restartFunctionByPulsarAdmin(Assignment assignment, String tenant, String namespace, String functionName)
```
what about
`restartFunctionUsingPulsarAdmin` ?

Not a big deal for me, just a suggestion

golden-yang replied:
I think your proposal is better. Thanks.

eolivelli reviewed (approved):
LGTM

freeznet reviewed (approved):
LGTM

golden-yang commented:
Thanks for suggest from @xche. I found that in thread model and k8s model, the restart code is very similar in both cases, but not exactly the same.  I have fixed this problem and added a unit test for the k8s scene.Please check again @xche .

sijie commented:
/pulsarbot run-failure-checks

golden-yang commented:
/pulsarbot run-failure-checks

golden-yang commented:
/pulsarbot run-failure-checks

golden-yang commented:
/pulsarbot run-failure-checks

golden-yang commented:
/pulsarbot run-failure-checks

jerrypeng commented on pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:
```diff
@@ -452,25 +444,43 @@ public synchronized void restartFunctionInstances(String tenant, String namespac
+    /**
+     * Restart the entire function or restart a single instance of the function
+     */
+    private void restartFunctionUsingPulsarAdmin(Assignment assignment, String tenant, String namespace,
```
I would recommend refactoring this method to the following: 

`
// restarts a particular instance of the function
private void internalRestartFunction(Assignment assignment, String tenant, String namespace, Integer instanceId)

// restarts the whole function
private void internalRestartFunction(Assignment assignment, String tenant, String namespace) {
  internalRestartFunction(assignment, tenant, namespace, null)
}
`

golden-yang replied:
Thank you for your advice.
I realized that in my previous implementation, using the boolean value `restartEntireFunction ` to control function behavior was absolutely bad code.

How about split it into two methods, 'internalrestartinstance' and 'internalrestartfunction', will it be more intuitive？
// restarts a particular instance of the function
`private void internalRestartInstance(Assignment assignment, String tenant, String namespace, Integer instanceId){}`

// restarts the whole function
`private void internalRestartFunction(Assignment assignment, String tenant, String namespace) {}`

Do you think it's necessary? Or just take your advice directly.
@jerrypeng

sijie replied:
@jerrypeng Do you want to check @golden-yang comment?

jerrypeng reviewed (changes requested):
@golden-yang thanks for contributing.  I have added a comment.

jerrypeng reviewed (commented):
Added a comment
